Using KNearestNeighbors to predict whether a person have a heart disease or not.

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import  confusion_matrix
from sklearn.metrics import  f1_score
from sklearn.metrics import  accuracy_score

In [11]:
dataset = pd.read_csv("E:\CSNotes\Machine learning\Datasets\heart.csv")

In [12]:
print(len(dataset))
print(dataset.head())

303
   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  


In [19]:
zero_not_accepted = ["sex","cp","fbs","restecg","exang","oldpeak","slope","ca"]

for column in zero_not_accepted:
    dataset[column] = dataset[column].replace(0, np.NaN)
    mean = int(dataset[column].mean(skipna = True))
    dataset[column] = dataset[column].replace(np.NaN, mean)
    

In [20]:
# split dataset
x = dataset.iloc[0:, 0:13]
y = dataset.iloc[:,13]
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state = 0, test_size = 0.2)

In [21]:
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)

In [22]:
#define the model
len(y)


303

In [23]:
import math
math.sqrt(len(y_test))

7.810249675906654

In [24]:
classifier= KNeighborsClassifier(n_neighbors = 7,p=2,metric='euclidean')

In [25]:
classifier.fit(x_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
           metric_params=None, n_jobs=1, n_neighbors=7, p=2,
           weights='uniform')

In [40]:
#predict the model
y_Pred = classifier.predict(x_test)

In [41]:
y_Pred

array([0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0], dtype=int64)

In [42]:
#evaluate the model
cm = confusion_matrix(y_test,y_Pred)
print(cm)

[[19  8]
 [ 4 30]]


In [43]:
print(f1_score(y_test, y_Pred))

0.833333333333


In [44]:
print(accuracy_score(y_test,y_Pred))

0.803278688525


This is the prediction of whether a person should take a test for heart disease. Accuracy of predicting the results is 80%.
And an f1 score of 83%.